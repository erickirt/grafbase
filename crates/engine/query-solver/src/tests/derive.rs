use crate::assert_solving_snapshots;

const SCHEMA: &str = r#"
    type Product
      @join__type(graph: EXT)
    {
      authorId: ID!
      commentIds: [ID!]!
      inventoryCountryId: ID!
      inventoryWarehouseId: ID!
      inventoriesKeys: [InventoryKeys!]!
      code: String!
      id: ID!
      author: User! @composite__derive(graph: EXT)
      comments: [Comment!]! @composite__derive(graph: EXT)
      inventory: Inventory! @composite__derive(graph: EXT)
      inventories: [Inventory!]! @composite__derive(graph: EXT)
    }

    type User
        @join__type(graph: EXT, key: "id", resolvable: false)
    {
      id: ID!
    }

    type Comment
        @join__type(graph: EXT, key: "id", resolvable: false)
    {
      id: ID!
    }

    type InventoryKeys @join__type(graph: EXT) {
        countryId: ID!
        warehouseId: ID!
    }

    type Inventory
        @join__type(graph: EXT, key: "countryId warehouseId", resolvable: false)
    {
        countryId: ID!
        warehouseId: ID!
    }

    type Query
    {
      products: [Product!]! @join__field(graph: EXT)
    }

    enum join__Graph
    {
    EXT @join__graph(name: "ext", url: "http://localhost:8080")
    }
"#;

#[test]
fn single_id() {
    assert_solving_snapshots!(
        "single_id",
        SCHEMA,
        r#"
        query {
            products {
                author {
                    id
                }
            }
        }
        "#
    );
}

#[test]
fn composite_keys() {
    assert_solving_snapshots!(
        "composite_keys",
        SCHEMA,
        r#"
        query {
            products {
                 inventory {
                    countryId
                    warehouseId
                }
            }
        }
        "#
    );
}

#[test]
fn batch_single_id() {
    assert_solving_snapshots!(
        "batch_single_id",
        SCHEMA,
        r#"
        query {
            products {
                comments {
                    id
                }
            }
        }
        "#
    );
}

#[test]
fn batch_composite_keys() {
    assert_solving_snapshots!(
        "batch_composite_keys",
        SCHEMA,
        r#"
        query {
            products {
                inventories {
                    countryId
                    warehouseId
                }
            }
        }
        "#
    );
}
