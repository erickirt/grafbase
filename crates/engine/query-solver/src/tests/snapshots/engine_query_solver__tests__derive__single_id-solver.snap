---
source: crates/engine/query-solver/src/tests/derive.rs
expression: "digraph {\n    0 [ label = \"root\", color=forestgreen ]\n    1 [ label = \"User.id\", style=dashed ]\n    2 [ label = \"Root#ext\", shape=parallelogram, color=dodgerblue, style=dashed ]\n    3 [ label = \"products#ext\", shape=box, color=dodgerblue, style=dashed ]\n    4 [ label = \"author#ext@derive\", shape=box, color=dodgerblue, style=dashed ]\n    5 [ label = \"id#ext@derive\", shape=box, color=dodgerblue, style=dashed ]\n    6 [ label=\"\", style=dashed]\n    0 -> 2 [ label = <<b>1</b>>, color=royalblue,fontcolor=royalblue,style=dashed ]\n    2 -> 3 [ label = \"\", color=royalblue,fontcolor=royalblue,style=dashed ]\n    3 -> 4 [ label = \"\", color=royalblue,fontcolor=royalblue,style=dashed ]\n    4 -> 5 [ label = \"\", color=royalblue,fontcolor=royalblue,style=dashed ]\n    5 -> 1 [ label = \"\", color=royalblue,fontcolor=royalblue,style=dashed ]\n    6 -> 0 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n}\n"
---
digraph {
    0 [ label = "root", steiner=1 ]
    1 [ label = "User.id", steiner=0 ]
    2 [ label = "Root#ext", steiner=0 ]
    3 [ label = "products#ext", steiner=0 ]
    4 [ label = "author#ext@derive", steiner=0 ]
    5 [ label = "id#ext@derive", steiner=0 ]
    6 [ label="", style=dashed]
    0 -> 2 [ cost=1, steiner=0]
    2 -> 3 [ cost=0, steiner=0]
    3 -> 4 [ cost=0, steiner=0]
    4 -> 5 [ cost=0, steiner=0]
    5 -> 1 [ cost=0, steiner=0]
    6 -> 0 [ cost=0, steiner=1]
}
